### PR TITLE
planner: use code-gen to generate CloneForPlanCache method for some operators

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #54057

Problem Summary: planner: use code-gen to generate CloneForPlanCache method for some operators

### What changed and how does it work?

planner: use code-gen to generate CloneForPlanCache method for some operators

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/plan_clone_test.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package core
```
Comment (hawkingrei) [inline pkg/planner/core/plan_clone_test.go:1]: Add lisence header.

Comment (qw4990) [reply pkg/planner/core/plan_clone_test.go:15]: added

**Review by hawkingrei — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

File: pkg/planner/core/physical_plans.go [hunk 2]
```diff
@@ -680,26 +680,26 @@ type PhysicalIndexScan struct {
 	// AccessCondition is used to calculate range.
 	AccessCondition []expression.Expression
 
-	Table      *model.TableInfo
-	Index      *model.IndexInfo
+	Table      *model.TableInfo `plan-cache-clone:"shallow"`
```
Comment (time-and-fate) [inline pkg/planner/core/physical_plans.go:683]: Should we add some comments on what do they mean and list all possible values?

File: pkg/planner/core/plan_clone_generator.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+import (
+	"bytes"
+	"fmt"
+	"go/format"
+	"reflect"
+)
+
+// genPlanCloneForPlanCacheCode generates CloneForPlanCache method for all physical plan nodes.
+// Using code-gen is safer than writing by hand, for example, if someone adds a new field to a struct,
+// the code-gen can update the Clone method correctly and automatically.
```
Comment (time-and-fate) [inline pkg/planner/core/plan_clone_generator.go:26]: Is there any instruction on how to update it automatically?

Comment (qw4990) [reply pkg/planner/core/plan_clone_generator.go:26]: I added more comments, PTAL

File: pkg/planner/core/physical_plans.go (hunk 2 above)
Comment (qw4990) [reply pkg/planner/core/physical_plans.go]: I added more comments on genPlanCloneForPlanCacheCode, PTAL

File: pkg/planner/core/physical_plans.go [hunk 4]
```diff
@@ -680,26 +680,26 @@ type PhysicalIndexScan struct {
...
 	// will be different. The schema of index scan will decode all columns of index but the TiDB only need some of them.
-	dataSourceSchema *expression.Schema
+	dataSourceSchema *expression.Schema `plan-cache-clone:"shallow"`
 
 	rangeInfo string
 
 	// The index scan may be on a partition.
 	physicalTableID int64
 
-	GenExprs map[model.TableItemID]expression.Expression `json:"-"`
```
Comment (time-and-fate) [inline pkg/planner/core/physical_plans.go:702]: Why is this tag removed?

**Review by time-and-fate — COMMENTED** [review]
Others LGTM.

**Pull request merged**